### ISSUE TITLE
Interaction of SIMD (VecElement) and closure can corrupt program

### ISSUE THREAD

**Issue by tkf** [JuliaLang/julia#30951]
Edit: reduced version here: https://github.com/JuliaLang/julia/issues/30951#issuecomment-462107012

---

I have code of the form `let <other bindings>, j = j; <let body>; end` where `j` is read only inside `<let body>`.  I'm doing this to avoid type instability due to closure #15276.  IIUC, this should not change the result of the computation.  Is it correct?  However, it changes the result in my code.

Here is how to reproduce:

```julia
]add https://github.com/tkf/SparseXX.jl#bd125934ce2dbb101407d93141fc74dc89c96792
]test SparseXX
```

passes the test (see: https://travis-ci.com/tkf/SparseXX.jl/builds/99578234) while

```julia
]add https://github.com/tkf/SparseXX.jl#2183a81c039da5464d825a8903cd935a780aa3cc
]test SparseXX
```

fails (see: https://travis-ci.com/tkf/SparseXX.jl/builds/99578239).  The change between those two revision is:

```diff
diff --git a/src/linalg.jl b/src/linalg.jl
index f48aa35..e334449 100644
--- a/src/linalg.jl
+++ b/src/linalg.jl
@@ -233,22 +233,22 @@ end
         @inbounds for col = 1:S1.n
             vaccs = map(triplets) do (_, S, _)
                 zero(Vec{N, eltype(S)})  # TODO: promote
             end

             nzr = nzrange(S1, col)
             simd_end = last(nzr) - N + 1
             j = first(nzr)
             while j <= simd_end
                 let lane = lane,
-                    idx = nzind[lane + j] #=,
-                    j = j =#
+                    idx = nzind[lane + j],
+                    j = j
                     vaccs = map(vaccs, nzvs, Xs) do vacc, nzv, X
                         @inbounds muladd(nzv[lane + j], X[idx], vacc)
                     end
                 end
                 j += N
             end

             accs = map(sum, vaccs)
             while j <= last(nzr)
                 accs = let j = j, idx = nzind[j]
```

I also noticed other changes such as shuffling the order of bindings also change the result (making previously failing test pass):

```diff
diff --git a/src/linalg.jl b/src/linalg.jl
index e334449..21c1713 100644
--- a/src/linalg.jl
+++ b/src/linalg.jl
@@ -239,8 +239,8 @@ end
             simd_end = last(nzr) - N + 1
             j = first(nzr)
             while j <= simd_end
-                let lane = lane,
-                    idx = nzind[lane + j],
+                let idx = nzind[lane + j],
+                    lane = lane,
                     j = j
                     vaccs = map(vaccs, nzvs, Xs) do vacc, nzv, X
                         @inbounds muladd(nzv[lane + j], X[idx], vacc)
```

This happens in Julia 1.0.3, 1.1.0, and 1.2-DEV (538fd8a2f7).

**Comment by tkf** [comment]
I find that I can workaround the issue by moving the code out of the function (https://github.com/tkf/SparseXX.jl/commit/6ff832f9b6773c8c8e4ebed2343d67f84fa2ffc0; https://travis-ci.com/tkf/SparseXX.jl/builds/99579723).  That is to say, this change fixes the test and type instability.  I'm noting it here in case someone else have a similar issue.

```diff
diff --git a/src/linalg.jl b/src/linalg.jl
index f48aa35..81eb202 100644
--- a/src/linalg.jl
+++ b/src/linalg.jl
@@ -237,17 +237,20 @@ end

             nzr = nzrange(S1, col)
             simd_end = last(nzr) - N + 1
             j = first(nzr)
             while j <= simd_end
+                #=
                 let lane = lane,
                     idx = nzind[lane + j] #=,
                     j = j =#
                     vaccs = map(vaccs, nzvs, Xs) do vacc, nzv, X
                         @inbounds muladd(nzv[lane + j], X[idx], vacc)
                     end
                 end
+                =#
+                vaccs = compute_vaccs(vaccs, nzvs, Xs, nzind[lane + j], lane + j)
                 j += N
             end

             accs = map(sum, vaccs)
             while j <= last(nzr)
@@ -267,5 +270,13 @@ end
             Y[col, k] += +(prods...)
         end
     end
     return Y
 end
+
+# This is a workaround for the possible bug in `let`:
+# https://github.com/JuliaLang/julia/issues/30951.  This fixes the
+# test _and_ makes the inference work.
+@inline compute_vaccs(vaccs, nzvs, Xs, idx, vj) =
+    map(vaccs, nzvs, Xs) do vacc, nzv, X
+        @inbounds muladd(nzv[vj], X[idx], vacc)
+    end
```

(This is not as efficient as I hoped since `compute_vaccs` allocates a lot of tuples which I hoped that Julia would eliminate.  But that's a different issue.)

**Comment by JeffBezanson** [comment]
Can you try to create a minimal self-contained example from this?

**Comment by tkf** [comment]
Yeah, I should do that.

**Comment by tkf** [comment]
A slightly more reduced version in https://gist.github.com/tkf/738d14c7c4cc61548686201aef751d88/cf9128bed2c02985ed50009107f7ffd24140d209.  It seems that [this closure](https://gist.github.com/tkf/738d14c7c4cc61548686201aef751d88/cf9128bed2c02985ed50009107f7ffd24140d209#file-bug-jl-L260-L267)

```julia
                vacc = (function(vacc, nzv, X)
                    nomask = Vec{4, Bool}(true)
                    as = simd_getindex(nzv, lane + j)
                    bs = vgather(Vec{4, Float64},
                                 pointer(X) + sizeof(Float64) * (idx - 1),
                                 nomask, Val{false})
                    return muladd(as, bs, vacc)
                end)(vaccs[1], nzvs[1], Xs[1])
```

is required to trigger the failure.  [Replacing this to a `let`](https://gist.github.com/tkf/738d14c7c4cc61548686201aef751d88/acff3c9f9f3df587966759937cff07b44a05775b#file-bug-jl-L260-L269) did not trigger the issue.

**Comment by tkf** [comment]
OK so I managed to get rid of the majority of the code including the ones calling llvmcall: https://gist.github.com/tkf/738d14c7c4cc61548686201aef751d88/caec32c746d1bdd110097a9965211d298719edeb

```julia
using Test

module SparseXX

const VE = Base.VecElement

struct Vec{N,T}
    elts::NTuple{N,VE{T}}
    @inline Vec{N,T}(elts::NTuple{N, VE{T}}) where {N,T} = new{N,T}(elts)
end

struct VecRange{N}
    i::Int
end
@inline Base.:+(idx::VecRange{N}, j::Integer) where N = VecRange{N}(idx.i + j)


function fmul_shared_simd!()
    X1 = nothing
    lane = VecRange{4}(0)
    j = 1
    # let idx = Vec{4, Int}((VE(0), VE(0), VE(0), VE(0))),  # this does not assert
    let idx = Vec{4, Int}((VE(1), VE(1), VE(1), VE(1))) #=,
        j = j =#
        # Uncommenting #= ... =# above avoids the segfault but the
        # test/assert still fails.
        (function()
            J = lane + j
            @show J.i           # required
            eq = J.i == j
            # @assert J.i == j
            @show X1            # required
            @show idx           # required
            return eq
        end)()
    end
end

end # module
fmul_shared_simd! = SparseXX.fmul_shared_simd!


@test fmul_shared_simd!()
```

However, it now triggers a _segfault_ with this code as-is (previously it was just passing the test).  Uncommenting `, j = j` fixes the segfault but the test still fails.

**Comment by tkf** [comment]
I just noticed that replacing `VecElement` with a normal `struct` fixes it.  So, maybe it's related to SIMD?

```diff
diff --git a/bug.jl b/bug.jl
index ef4bf3f..eb654c1 100644
--- a/bug.jl
+++ b/bug.jl
@@ -2,7 +2,9 @@ using Test

 module SparseXX

-const VE = Base.VecElement
+struct VE{T}
+    value::T
+end

 struct Vec{N,T}
     elts::NTuple{N,T}
```

**Comment by tkf** [comment]
Bump. I still can reproduce the bug https://github.com/JuliaLang/julia/issues/30951#issuecomment-462107012 with current `master` 1.2.0-DEV.669 (4671132ee1).

**Comment by KristofferC** [comment]
Seems to work now? cc @tkf

**Comment by tkf** [comment]
This seems to be fixed in 1.4 (while it's reproducible 1.3).
